File: Hashtable_Flutter/lib/presentation/view/HashtableDrawComponent.dart
import 'package:demo/model/Table.dart' as t;
import 'package:flutter/material.dart';

import '../../model/Node.dart';

class HashtableDrawComponent extends CustomPainter {
  int _index = 0;

  Map<String, List<double>> _nodeMap = {};

  t.Table _model = t.Table(10);

  HashtableDrawComponent();

  t.Table getModel() => _model;

  t.Table setModel(t.Table model) => _model = model;

  void setIndex(index) => _index = _index;

  Map<String, List<double>> getMap() => _nodeMap;

  void setNodeMap(Map<String, List<double>> nodeMap) => _nodeMap = nodeMap;

  void drawNode(
      Canvas canvas, double nodeX, double nodeY, double caseSize, Node n) {
    Paint linePaint = Paint()
      ..color = Colors.black
      ..strokeWidth = 3;

    canvas.drawLine(
      Offset(nodeX - 40, nodeY + 30),
      Offset(nodeX, nodeY + 30),
      linePaint,
    );

    Rect nodeRect = Rect.fromLTRB(
      nodeX,
      nodeY,
      nodeX + caseSize - 20,
      nodeY + caseSize - 40,
    );

    Paint nodePaint = Paint()
      ..color = Colors.white
      ..style = PaintingStyle.fill;
    canvas.drawRect(nodeRect, nodePaint);
    final strokePaint = Paint()
      ..color = Colors.black
      ..style = PaintingStyle.stroke
      ..strokeWidth = 2;
    canvas.drawRect(nodeRect, strokePaint);

    TextSpan textSpan = TextSpan(
      text: n.value.toString(),
      style: const TextStyle(color: Colors.black),
    );
    TextPainter textPainter = TextPainter(
      text: textSpan,
      textDirection: TextDirection.ltr,
    );
    textPainter.layout();
    double textX = nodeX + 10;
    double textY = nodeY + 20;
    Offset textOffset = Offset(textX, textY);
    textPainter.paint(canvas, textOffset);
    Paint connectionLinePaint = Paint()
      ..color = Colors.black
      ..strokeWidth = 3;

    canvas.drawLine(
      Offset(nodeX + caseSize - 20, nodeY + 30),
      Offset(nodeX + caseSize + 20, nodeY + 30),
      connectionLinePaint,
    );
    _nodeMap[n.value.toString()] = [
      nodeX,
      nodeY,
      nodeX + caseSize - 20,
      nodeY + caseSize - 40
    ];

  }

  void drawMsalha(Canvas canvas, double nodeX, double lineY) {
    Paint linePaint = Paint()
      ..color = Colors.black
      ..strokeWidth = 3;

    canvas.drawLine(
      Offset(nodeX, lineY + 30),
      Offset(nodeX, lineY - 20),
      linePaint,
    );
    Paint line1 = Paint()
      ..color = Colors.black
      ..strokeWidth = 3;

    canvas.drawLine(
      Offset(nodeX, lineY - 5),
      Offset(nodeX + 10, lineY - 15),
      line1,
    );
    Paint line2 = Paint()
      ..color = Colors.black
      ..strokeWidth = 3;

    canvas.drawLine(
      Offset(nodeX, lineY + 5),
      Offset(nodeX + 10, lineY - 5),
      line2,
    );
    Paint line3 = Paint()
      ..color = Colors.black
      ..strokeWidth = 3;

    canvas.drawLine(
      Offset(nodeX, lineY + 15),
      Offset(nodeX + 10, lineY + 5),
      line3,
    );
    Paint line4 = Paint()
      ..color = Colors.black
      ..strokeWidth = 3;

    canvas.drawLine(
      Offset(nodeX, lineY + 25),
      Offset(nodeX + 10, lineY + 15),
      line4,
    );
  }

  @override
  void paint(Canvas canvas, Size size) {
    int tableSize = _model.getNodes().length;
    double caseSize = 100;
    double tableHeight = caseSize * tableSize;
    double startX = (size.width / 2) - 250;
    double startY = ((size.height - tableHeight) / 2);
    for (int i = 0; i < tableSize; i++) {
      final paint = Paint()
        ..color = Colors.white
        ..style = PaintingStyle.fill;
      double y = startY + i * caseSize;
      Rect rect = Rect.fromLTRB(startX, y, startX + caseSize, y + caseSize);

      canvas.drawRect(rect, paint);
      final strokePaint = Paint()
        ..color = Colors.black
        ..style = PaintingStyle.stroke
        ..strokeWidth = 2;
      canvas.drawRect(rect, strokePaint);
      double nodeX = startX + caseSize + 40;
      Node? n = _model.getNodes()[i];
      while (n != null) {
        double nodeY = y + caseSize - 75;
        drawNode(canvas, nodeX, nodeY, caseSize, n);
        n = n.getNext();
        nodeX += caseSize + 20;
      }
      double lineY = y + caseSize / 2;
      drawMsalha(canvas, nodeX, lineY);
    }
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) {
    return true;
  }
}

File: Hashtable-JavaFX/src/main/java/com/example/demo2/presentation/view/HashtableDrawComponent.java
package com.example.demo2.presentation.view;

import com.example.demo2.model.Node;
import com.example.demo2.model.Table;
import javafx.animation.FillTransition;
import javafx.animation.KeyFrame;
import javafx.animation.Timeline;
import javafx.scene.input.MouseButton;
import javafx.scene.layout.Pane;
import javafx.scene.paint.Color;
import javafx.scene.shape.Line;
import javafx.scene.shape.Rectangle;
import javafx.scene.text.Text;
import javafx.util.Duration;

public class HashtableDrawComponent {
    private String delName;
    private Boolean delTest = false;
    private boolean isAdded = false;
    private int index;
    private boolean confirmed = false;
    private Table model = new Table(10);

    public HashtableDrawComponent() {

    }

    public void setConfirmed(boolean confirmed) {
        this.confirmed = confirmed;
    }

    public Boolean getDelTest() {
        return delTest;
    }

    public String getDelName() {
        return delName;
    }

    public Table getModel() {

        return model;
    }

    public void setModel(Table model) {
        this.model = model;
//        Platform.runLater(this::paintComponent);
    }

    public void setAdded(boolean added) {
        this.isAdded = added;
    }

    public void setIndex(int index) {
        this.index = index;
    }

    public Rectangle drawNode(Pane pane, Node n, int nodeX, int nodeY, int caseSize) {
        Rectangle nodeRect = new Rectangle(nodeX, nodeY, caseSize - 20, caseSize - 40);
        nodeRect.setFill(Color.WHITE);
        nodeRect.setStroke(Color.BLACK);
        nodeRect.setStrokeWidth(1);
        pane.getChildren().add(nodeRect);

        Text nodeText = new Text(n.getValue());
        nodeText.setX(nodeX + 15);
        nodeText.setY(nodeY + 30);
        pane.getChildren().add(nodeText);

        Line line = new Line(nodeX, nodeY + 25, nodeX - 40, nodeY + 25);
        line.setStroke(Color.BLACK);
        line.setStrokeWidth(1);
        pane.getChildren().add(line);

        Line connectionLine = new Line(nodeX + caseSize - 17, nodeY + 25, nodeX + caseSize + 20, nodeY + 25);
        connectionLine.setStroke(Color.BLACK);
        connectionLine.setStrokeWidth(1);
        pane.getChildren().add(connectionLine);
        nodeRect.setOnMouseClicked(event -> {
            if (event.getButton() == MouseButton.PRIMARY) {
                nodeRect.setFill(Color.GRAY);
                delTest = true;
                delName = n.getValue();
                System.out.println(delName);
                if (confirmed) {
                    isAdded = false;

                } else {
                    nodeRect.setFill(Color.WHITE);

                }


            }
        });


        nodeRect.setFill(Color.WHITE);
        delTest = false;
        return nodeRect;
    }

    public void drawMsalha(Pane pane, int nodeX, int lineY) {
        Line line1 = new Line(nodeX, lineY + 20, nodeX, lineY - 20);
        line1.setStroke(Color.BLACK);
        line1.setStrokeWidth(1);
        pane.getChildren().add(line1);

        Line line2 = new Line(nodeX, lineY + 10, nodeX + 12, lineY - 2);
        line2.setStroke(Color.BLACK);
        line2.setStrokeWidth(1);
        pane.getChildren().add(line2);

        Line line3 = new Line(nodeX, lineY, nodeX + 12, lineY - 12);
        line3.setStroke(Color.BLACK);
        line3.setStrokeWidth(1);
        pane.getChildren().add(line3);
        Line line4 = new Line(nodeX, lineY - 10, nodeX + 12, lineY - 22);
        line4.setStroke(Color.BLACK);
        line4.setStrokeWidth(1);
        pane.getChildren().add(line4);

    }

    public void animateGreen(Pane pane, int startX, Rectangle cell, FillTransition fillG, int y, int i) {
        Text text = new Text(String.valueOf(i));
        Timeline timeline = new Timeline(

                new KeyFrame(Duration.seconds((i + 1)), event -> {
                    text.setX(startX - 50);
                    text.setY(y + 50);
                    pane.getChildren().add(text);
                    cell.setFill(Color.GREEN);
                    fillG.setFromValue(Color.GREEN);
                    fillG.setToValue(Color.WHITE);
                    fillG.setShape(cell);
                    fillG.play();


                })


        );
        timeline.play();
    }

    public void animateOrange(int i, int j, Rectangle rec) {
        Timeline timeline = new Timeline(

                new KeyFrame(Duration.seconds((j + i + 1)), event -> {
                    FillTransition fillO = new FillTransition();
                    rec.setFill(Color.ORANGE);
                    fillO.setFromValue(Color.ORANGE);
                    fillO.setToValue(Color.WHITE);
                    fillO.setShape(rec);
                    fillO.play();

                })


        );
        timeline.play();
    }

    public Pane paintComponent() {
        Pane pane = new Pane();

        int tableSize = model.getNodes().length;
        int caseSize = 100;
        int tableHeight = caseSize * tableSize;

        int startX = (int) (pane.getWidth() - caseSize * 2 - 100) / 2 + 500;
        int startY = (int) (pane.getHeight() - tableHeight) / 2 + 300;
        FillTransition fillG = new FillTransition();

        for (int i = 0; i < tableSize; i++) {
            System.out.println("case : " + i);
            int y = startY + i * caseSize;
            Rectangle cell = new Rectangle(startX, y, caseSize, caseSize);
            cell.setFill(Color.WHITE);
            boolean test = false;
            cell.setStroke(Color.BLACK);
            cell.setStrokeWidth(3);
            pane.getChildren().add(cell);
            Node n = model.getNodes()[i];
            int nodeX = startX + caseSize + 40;
            Rectangle node;
            int j = 0;

            while (n != null) {
                test = true;
                int nodeY = y + caseSize - 75;
                node = drawNode(pane, n, nodeX, nodeY, caseSize);
                nodeX += caseSize + 20;
                n = n.getNext();
                j = j + 1;

                if (index == i && isAdded) {
                    animateOrange(i, j, node);

                }
            }


            if (index >= i && isAdded) {
                animateGreen(pane, startX, cell, fillG, y, i);

            }
            if (test) {
                j = j + 1;
                int lineY = y + caseSize / 2;
                int finalNodeX = nodeX;
                if (index == i) {
                    Timeline timeline;
                    if (confirmed) {
                        timeline = new Timeline(

                                new KeyFrame(Duration.seconds((j * 0.75)), event ->
                                        drawMsalha(pane, finalNodeX, lineY))

                        );
                        confirmed = false;
                    } else {
                        timeline = new Timeline(

                                new KeyFrame(Duration.seconds((i + j + 1)), event ->
                                        drawMsalha(pane, finalNodeX, lineY))

                        );

                    }
                    timeline.play();

                } else {
                    drawMsalha(pane, finalNodeX, lineY);
                }

            }


        }
        isAdded = false;
        return pane;
    }

}
